Draw scheduled line when a vehicle is scheduled to be on a different trip

skyqrose opened the pull request:
Asana Task: [Calculate scheduled time based or block instead of trip](https://app.asana.com/0/1112935048846093/1125606672346265)

<img width="209" alt="Screen Shot 2019-07-01 at 10 06 27" src="https://user-images.githubusercontent.com/23065557/60448510-c2426480-9bf3-11e9-8a95-e4f8110cc12c.png">

- If a block isn't scheduled to have started yet: draw to the start of the first trip
- If a block is scheduled to have finished: draw to the end of the last trip
- If it's scheduled to be in the middle of a layover: draw to the end of the previous trip
- If it's scheduled to be  in the middle of a trip: draw to the next timpeoint in that trip

I had some trouble finding real examples of the bus being so behind it was scheduled to be on a different trip. More commonly, the buses with bad data in the shame corner have a really distracting line all the way across the screen. I think that's okay for now since we're planning on fixing that bug anyway.

<img width="242" alt="Screen Shot 2019-07-01 at 09 55 58" src="https://user-images.githubusercontent.com/23065557/60448900-865bcf00-9bf4-11e9-98d2-d65b3ec4e3cd.png">

skyqrose commented:
They were listed twice. I just removed the second.
(Sorry this PR is so big.)

arkadyan commented:
@skyqrose Oh, oops! 👍

arkadyan commented on lib/gtfs/block.ex:
```diff
@@ -0,0 +1,26 @@
+  @type by_id :: %{{id(), Service.id()} => t()}
+
+  @spec group_trips_by_block([Trip.t()]) :: by_id()
+  def group_trips_by_block(trips) do
```
Another variation to consider (sorry for the rambling as I think this over):

    @type block_id:: String.t()
    @type service_id :: String.t()
    @type id :: {block_id(), service_id()}
    @type by_id :: %{id() => t()}

(The Service module could go away as well)

skyqrose replied:
I guess the confusing part is that blocks aren't uniquely determined by their block_id. They're determined by the `block_id, service_id` pair. So which should be called `id`? I think, since `block_id` is an existing concept, we should keep it, and avoid making `Block.id()` refer to a different concept.

Pulling the map type and the get function into `Block` was my attempt to isolate the outside world from this confusion, but it doesn't remove the confusion from inside the `Block` module.

What do you think about just adding a `@doc` explaining why you need a service id to get a block?

arkadyan replied:
I think that the unique determiner for a module should be the ID, which makes me think that `@type id :: {block_id(), service_id()}` is the important part. What if we named this module something else other than block, so we have `block_id` and `service_id` but those don't need to reference other modules?

arkadyan replied:
`key` is a better name for the `{block_id(), service_id()}` tuple we just came up with.

arkadyan commented on lib/realtime/vehicle.ex:
```diff
@@ -199,31 +204,87 @@ defmodule Realtime.Vehicle do
-         end) do
-      nil ->
-        nil
+  def scheduled_location(block, now) do
```
Could you just fetch and call `now_fn` inside `scheduled_location` so you don't need to pass it in?

skyqrose replied:
When unit testing, it's easier to pass in the values to a pure function, rather than having to set the environment for every test that uses a different input.
E.g.
```ex
block = ...
now = 1_546_362_000 # 2019-01-01 12:00:00 EST
assert Vehicle.scheduled_location(block, now) == ...
```
(https://github.com/mbta/skate/pull/84/files#diff-efc218bd9fe2e859af974b97d7770845R442)

as opposed to
```ex
block = ...
now = ...
Application.put_env(:realtime, :block_fn, fn _ -> block end)
Application.put_env(:realtime, :now_fn, fn -> now end)
assert Vehicle.scheduled_location(block_id, service_id) == ...
```

I still have to do that when testing the higher level `from_vehicle_position_and_trip_update`, but I'm not testing all the edge cases there.

arkadyan commented on lib/realtime/vehicle.ex:
```diff
@@ -199,31 +204,87 @@ defmodule Realtime.Vehicle do
+    block_end = List.last(List.last(block).stop_times).time
+    now_time_of_day = Util.Time.nearest_time_of_day_for_timestamp(now, block_start)
+
+    trip =
```
Could this be broken out into a private function? Same with `timepoint_status` assignment below. This function is pretty long, so I think that any way of condensing what is happening might be helpful.

arkadyan commented on assets/src/hooks/useVehicles.ts:
```diff
@@ -229,10 +237,12 @@ const vehicleFromData = (vehicleData: VehicleData): Vehicle => ({
-    vehicleData.scheduled_timepoint_status
-  ),
+  timepointStatus:
+    vehicleData.timepoint_status &&
```
If only we had matching in JS…

skyqrose replied:
😢I miss case expressions and the [`Maybe` type](https://package.elm-lang.org/packages/elm/core/1.0.2/Maybe#map)

skyqrose commented on lib/gtfs/block.ex:
```diff
@@ -0,0 +1,26 @@
+  end
+
+  @spec get(by_id(), id(), Service.id()) :: t() | nil
+  def get(by_id, block_id, service_id) do
```
Rather than testing this function independently, I've added tests that check that `group_trips_by_block` and `get` work together when used in sequence so I don't have to make a bunch of maps that depend on implementation details.

skyqrose commented on lib/gtfs/data.ex:
```diff
@@ -60,6 +63,11 @@ defmodule Gtfs.Data do
   def trip(%__MODULE__{trips: trips}, trip_id), do: trips[trip_id]
 
+  @spec block(t(), Block.id(), Service.id()) :: Block.t() | nil
+  def block(%__MODULE__{blocks: blocks}, block_id, service_id) do
```
Hm. It'd be nice to be able to work with a single `key` argument, but I'd rather avoid having to construct the `key` tuple outside of `Block`.

skyqrose commented:
Okay. Ready for another review.

You made a comment today about buses being scheduled to be on other routes that made me realize there was a bug where we'd incorrectly interpret the `scheduled_location` field in that case, so I added a check that we're scheduled to be on the same route, and removed the field if we're scheduled to be on a different route.
